### DISCUSSION TITLE
Part2 convergence

### DISCUSSION THREAD

**Discussion by Julian-Wyatt** [wenceorg/comp4187#96] [Q&A]
I'm struggling to get a convergence on the part 2a code. I had fixed my stencil following your previous discussion #91 .
I initially thought it was alright as it converged, produced a scaled error (multiplied by all degrees of freedom `HxHyHzHt`) of 1e-6 and a max value of ~1.27 (compared to exact of 1).

However I had missed multiplying the inside of the cos function by pi, which has since removed all convergence. 

Do you have any suggestions for directions to fix this?

**Comment by annereinarz** [comment]
Is the Ht there as part of the Euler-step or are you using that to scale both sides? The scaling of the operator should only be done using the spatial components.

**Reply by Julian-Wyatt** [reply]
I've only used Ht for the Euler stepping and the error scaling (`np.linalg.norm(u.array - exact) * (Hx * Hy * Hz * Ht)`) - to remove the margin for error, I've tried the calculation with and without the volume scaling (multiply rhs and lhs by `HxHyHz`). Worth noting, that with the volume scaling performs closer to the expected result, but the derivative is still increasing, so never ends up on a result.

**Reply by annereinarz** [reply]
That sounds reasonable. I think I still only have the same answer as before: If this happens both with LU and multigrid the error is somewhere in the matrix setup or rhs. If you've missed a factor of pi perhaps you've missed one in the exact solution as well?

**Reply by shadowbourne** [reply]
I believe I am having similar issues with convergence for question 2 as I have been struggling for days to make my solution converge rather than diverge. (But nothing to do with missing Pi's afaik)

I have derived my stencil over and over and cannot work out why my solution does not converge :(

Has anyone managed to solve q2a??

**Reply by annereinarz** [reply]
I guess one thing to watch out for is the boundary conditions, these are zero so make sure you're not adding anything on those nodes in the iteration. I think the 1D test case by default sets u[0] = 0 by placing an identity matrix there... This won't work for time-stepping.

You can test the matrix without the Euler time-stepping, by directly solving the steady state equation with e.g. LU or multigrid as suggested above.

The other thing to watch out for is the balance between Ht and Hx, Hy and Hz. Assuming that Hx ~ Hy ~ Hz for explicit Euler we want Ht ~ Hx^2. For part 2a k=1, but otherwise that also depends on k.

**Reply by Julian-Wyatt** [reply]
To clarify, the stencil should be `- \nabla \cdot K(x, y, z) \nabla u`. Obviously, in the EE case you move that to the other side of the equation, so in fact you then add the negative stencil with b?
I doubt the above is the problem as neither works

I've tried the LU solver to no avail, but this could also be from inserting vs adding stencil values.

When you say 2a k=1, I didn't think there was a K parameter for question 2? (outside of the cos(pi y))

**Reply by annereinarz** [reply]
I'm not sure what you mean by "negative stencil with b". The system I meant was ` - \nabla \cdot K(x, y, z) \nabla u = f(x,y,z)`, where you're probably writing` -(k' d_y+k \Delta) u = f`

Sorry, that was badly phrased, in 2a |k|<=1 so we can ignore it in the time-step restriction.

**Reply by annereinarz** [reply]
For explicit Euler you presumably rewrite as:
d_t u = f +  \nabla \cdot K(x, y, z) \nabla u

**Reply by Julian-Wyatt** [reply]
So in the question, you've asked to implement the negative stencil: 
![Screenshot 2022-04-27 at 11 26 07](https://user-images.githubusercontent.com/44175667/165498257-120c6b7a-7281-4320-a871-e14465e0e697.png)

So, exactly when you rewrite as `d_t u = f + \nabla \cdot K(x, y, z) \nabla u`, you'd have` - - ( \nabla \cdot K(x, y, z) \nabla u)`, which are equivalent.

Which does mean my stencil looks identical to the: dudt = (k' d_y+k \Delta) u + f.

Therefore, I doubt this is where the error is coming from.

**Reply by annereinarz** [reply]
Makes sense. And the issues with testing the steady state arise from having implemented only the multiplication of the matrix with a vector (`mult()`) not the sparse matrix (`as_sparse_matrix()`)?

**Reply by Julian-Wyatt** [reply]
I've implemented the sparse matrix function, I'm just less confident it is correct due to the add values vs insert values (logically they're all add). 
While it does "solve" with LU, it produces an error of 2785701650847069.5, when scaling it not applied to either side which does imply that, like you say, the stencil or rhs are in correct.

**Reply by shadowbourne** [reply]
I have also implemented the sparse matrix function, which I attempt to solve `-(k' d_y+ k (d_yy + d_zz + d_xx)) u = f` with.
However, the solver does not converge for most grid sizes, and never produces a reasonable result.
Additionally, when using mult to run explicit euler, (having different boundary conditions of course) my solution still diverges.

However I am fairly confident with the maths I have done and the resulting stencil, as I have gone through at many a time (however, something is obviously wrong with my stencil, I just don't know where to look to find out what :( )

**Reply by annereinarz** [reply]
What I would do is try to find some tests that isolate the issue. So in this case, first I would check if the solution without the Euler and with are identical (to a certain tolerance). If not then at least one of the two functions isn't implementing what you have derived on paper. 

Next I'd probably try to work out which part of the stencil is causing issues. So if you just let k=1, you'll be left only with d_yy + d_zz + d_xx and you can check this. The exact solution in that case remains the same if you change the rhs to 3π^2*sin(πx)sin(πy)sin(πz)

**Reply by Julian-Wyatt** [reply]
This is very similar to what I had before - so this does now converge on small grids, but on larger grids it doesn't:
<img width="643" alt="Screenshot 2022-05-01 at 18 57 07" src="https://user-images.githubusercontent.com/44175667/166158346-c0e4c8c9-acbe-426c-95e1-8ce2d7c45093.png">

Also when it converges the maximum value is `1.2337004766018866`, where it should be 1. 

Could this be solely down to the convergence order of the stencil?

**Reply by Julian-Wyatt** [reply]
I've since realised this was due to the number of iterations in the EE, as they increase exponentially, such that a 5x5x5 grid requires 12K iters and it does get closer to 0
